### PR TITLE
Prefetch code cleanup

### PR DESCRIPTION
**Pull request by sanoursa**
In preparation for optimizing noop prefetches, I've done some cleanup in the prefetch/fastfetch code. These are all intended to be non-functional changes.

The biggest purpose of this PR is to add the comments in `BlobPrefetcher` explaining how the prefetch pipeline stages interact with each other. 

The rest of it is:
* Moved fastfetch-specific classes into `FastFetch.csproj`
* Moved some classes out of the prefetch namespace because they are used by non-prefetch code
* Deleted unused classes
* Renamed "Job" to "PrefetchStage"

### CODE REVIEW

File: GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs [hunk 1]
```diff
@@ -229,38 +229,83 @@ public virtual void Prefetch(string branchOrCommit, bool isBranch)
...
             {
-                blobEnumerator.PerformDiff(previousCommit, commitToFetch);
-                this.HasFailures |= blobEnumerator.HasFailures;
+                diff.PerformDiff(previousCommit, commitToFetch);
+                this.HasFailures |= diff.HasFailures;
             };
 
             if (readFilesAfterDownload)
             {
                 // Call synchronously to ensure that blobEnumerator.FileAddOperations
```
Comment (kewillford) [inline GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs]: diff.FileAddOperations?

File: GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs [hunk 2]
```diff
@@ -229,38 +229,83 @@ public virtual void Prefetch(string branchOrCommit, bool isBranch)
...
 
             ThreadStart performDiff = () =>
             {
-                blobEnumerator.PerformDiff(previousCommit, commitToFetch);
-                this.HasFailures |= blobEnumerator.HasFailures;
+                diff.PerformDiff(previousCommit, commitToFetch);
+                this.HasFailures |= diff.HasFailures;
             };
 
             if (readFilesAfterDownload)
```
Comment (kewillford) [inline GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs]: Rename to hydrateFilesAfterDownload?

File: GVFS/GVFS.Common/Git/HashingStream.cs [hunk 3]
```diff
@@ -2,7 +2,7 @@
 using System.IO;
 using System.Security.Cryptography;
 
-namespace GVFS.Common.Prefetch
+namespace GVFS.Common
```
Comment (sanoursa) [inline GVFS/GVFS.Common/Git/HashingStream.cs]: Add `.Git`

File: GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs [hunk 4]
```diff
@@ -229,38 +229,83 @@ public virtual void Prefetch(string branchOrCommit, bool isBranch)
...
+            DiffHelper diff = new DiffHelper(this.Tracer, this.Enlistment, this.FileList, this.FolderList, includeSymLinks: false);
+
+            // blobFinder 
+            //  Inputs:
+            //      * requiredBlobs (in param): Blob ids from output of `diff`
+            //  Outputs:
+            //      * availableBlobs (out param): Locally available blob ids (shared between `blobFinder`, `downloader`, and `packIndexer`, all add blob ids to the list as they are locally available)
+            //      * MissingBlobs (property): Blob ids that are missing and need to be downloaded
+            //      * AvailableBlobs (property): Same as availableBlobs
+            FindMissingBlobsStage blobFinder = new FindMissingBlobsStage(this.SearchThreadCount, diff.RequiredBlobs, availableBlobs, this.Tracer, this.Enlistment);
```
Comment (kewillford) [inline GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs]: Rename to missingBlobsFinder?

Comment (sanoursa) [reply GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs]: It's actually finding missing _and_ available blobs, so the class is the one that's misnamed. I'll fix it up.

File: GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs [hunk 5]
```diff
@@ -229,38 +229,83 @@ public virtual void Prefetch(string branchOrCommit, bool isBranch)
...
+            //      * availableBlobs (out param): Loose objects that have completed downloading (shared between `blobFinder`, `downloader`, and `packIndexer`, all add blob ids to the list as they are locally available)
+            //      * AvailableObjects (property): Same as availableBlobs
+            //      * AvailablePacks (property): Packfiles that have completed downloading
+            BatchObjectDownloadStage downloader = new BatchObjectDownloadStage(this.DownloadThreadCount, this.ChunkSize, blobFinder.MissingBlobs, availableBlobs, this.Tracer, this.Enlistment, this.ObjectRequestor, this.GitObjects);
+
+            // packIndexer
+            //  Inputs:
+            //      * availablePacks (in param): Packfiles that have completed downloading from output of `downloader`
+            //  Outputs:
+            //      * availableBlobs (out param): Blobs that have completed downloading and indexing ((shared between `blobFinder`, `downloader`, and `packIndexer`, all add blob ids to the list as they are locally available)
```
Comment (jamill) [inline GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs]: typo - double paren - `((shared...`

File: GVFS/GVFS.Common/Prefetch/Pipeline/FindMissingBlobsJob.cs [hunk 6]
```diff
@@ -3,14 +3,14 @@
...
 using System.Threading;
 
-namespace GVFS.Common.Prefetch.Jobs
+namespace GVFS.Common.Prefetch.Pipeline
 {
     /// <summary>
     /// Takes in search requests, searches each tree as requested, outputs blocks of missing blob shas.
     /// </summary>
-    public class FindMissingBlobsJob : Job
+    public class FindMissingBlobsStage : PrefetchPipelineStage
```
Comment (sanoursa) [inline GVFS/GVFS.Common/Prefetch/Pipeline/FindMissingBlobsJob.cs]: Rename the file to match the new class

File: GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs [hunk 7]
```diff
@@ -229,38 +229,83 @@ public virtual void Prefetch(string branchOrCommit, bool isBranch)
                 }
             }
 
-            DiffHelper blobEnumerator = new DiffHelper(this.Tracer, this.Enlistment, this.FileList, this.FolderList, includeSymLinks: false);
```
Comment (sanoursa) [inline GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs]: Looks like line 214 had an earlier attempt to document the pipeline, but it is very sparse and also now in the wrong place. Remove it.

Comment (sanoursa) [reply GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs]: Will be handled in my next PR

**Pull request merged**